In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)


In [2]:

# List of file names
files = ['weekly_patterns_2018_sample.csv.zip', 
         'weekly_patterns_2019_sample.csv.zip',
         'weekly_patterns_2020_sample.csv.zip', 
         'weekly_patterns_2021_sample.csv.zip',
         'weekly_patterns_2022_sample.csv.zip']


total_obs = 0
for file in files:
    # Read the data
    data = pd.read_csv(file, compression='zip')
    total_obs += len(data)
print(total_obs)
data.dtypes
chilis_obs = 0
chilis_df = pd.DataFrame() #initializing an empty df

for file in files:
    # Read the data
    data = pd.read_csv(file, compression='zip')
    
    # Filter the data for Chili's Grill & Bar
    data = data.loc[data['brands'].isin(["Chili's Grill & Bar"])]
    
    # Concatenate the data
    chilis_df = pd.concat([chilis_df, data], ignore_index = True)
    
    chilis_obs = len(chilis_df)
    
print(chilis_obs)
from datetime import datetime, date, timedelta
chilis_df['date_range_start'] = chilis_df['date_range_start'].str[:10]
chilis_df['date_range_end'] = chilis_df['date_range_end'].str[:10]
chilis_df.head(5)

chilis_df['year'] = chilis_df['date_range_start'].str[:4]
chilis_df['date_range_start'] = pd.to_datetime(chilis_df
                                               ['date_range_start'])
chilis_df['date_range_end'] =  pd.to_datetime(chilis_df
                                              ['date_range_end'])
# Split the variable visits_by_day into new variables
chilis_df[['dailyvisits0', 'dailyvisits1', 'dailyvisits2'
           , 'dailyvisits3',
    'dailyvisits4', 'dailyvisits5', 
           'dailyvisits6']] = chilis_df['visits_by_day'].str.split(',', expand=True)
chilis_df['dailyvisits0'] = chilis_df['dailyvisits0'].str[1:3]
chilis_df['dailyvisits6'] = chilis_df['dailyvisits6'].str[0:2]

chilis_df['dailyvisits0'] = chilis_df['dailyvisits0'].astype(float)
chilis_df['dailyvisits1'] = chilis_df['dailyvisits1'].astype(float)
chilis_df['dailyvisits2'] = chilis_df['dailyvisits2'].astype(float)
chilis_df['dailyvisits3'] = chilis_df['dailyvisits3'].astype(float)
chilis_df['dailyvisits4'] = chilis_df['dailyvisits4'].astype(float)
chilis_df['dailyvisits5'] = chilis_df['dailyvisits5'].astype(float)

chilis_df['dailyvisits6'] = chilis_df['dailyvisits6'].str.replace(']', '')

chilis_df['dailyvisits6'] = chilis_df['dailyvisits6'].astype(float)

chilis_df['id'] = chilis_df.index
chilis_df['index'] = chilis_df.index
long_chilis_df = pd.wide_to_long(chilis_df,
                                 ['dailyvisits'
                                 ], i = 'index', j = 'day', sep = '')
long_chilis_df['dailyvisits'] = long_chilis_df['dailyvisits'].astype(int)
agg = long_chilis_df.groupby(['year']).agg(
{'dailyvisits': 'mean' })

long_chilis_df = long_chilis_df.reset_index()

long_chilis_df['date'] = long_chilis_df['date_range_start'
                                       ] + pd.to_timedelta(long_chilis_df['day'],
                                                           unit='D')

long_chilis_df['dayoftheweek'] = long_chilis_df['date'].dt.day_name()
threshold = 34 #I chose this threshold because 25% of the data will be greater than it: 
                #this gets the top quartile

long_chilis_df['manyvisits'] = np.where(long_chilis_df['dailyvisits'] > threshold, 1, 0)
table = pd.crosstab(long_chilis_df['region'] , long_chilis_df['manyvisits'] )
tablepercent = pd.crosstab(long_chilis_df['region'] , long_chilis_df['manyvisits'] ,
                           normalize = 'index')*100
#same table as above but with percentages instead of total number counts
#Done to control for state size and see which states have best performing resturants
long_chilis_df['core_biz_area'] = np.where((long_chilis_df['region'] == 'AL')
                                           | (long_chilis_df['region'] =='AR')
                                           | (long_chilis_df['region'] =='FL') 
                                           | (long_chilis_df['region'] =='GA')
                                           |(long_chilis_df['region'] == 'KS') 
                                           | (long_chilis_df['region'] =='LA')
                                           |(long_chilis_df['region'] == 'MI') 
                                           |(long_chilis_df['region'] == 'MO') 
                                           |(long_chilis_df['region'] == 'MS') 
                                           | (long_chilis_df['region'] =='NC') 
                                           |(long_chilis_df['region'] =='OK') 
                                           |(long_chilis_df['region'] == 'TX') 
                                           |(long_chilis_df['region'] == 'WV'), 1, 0)

long_chilis_df.head()

core_agg = long_chilis_df.groupby(['core_biz_area']).agg(
{'dailyvisits': 'mean' })

core_agg.head()

6462235
56242


dailyvisits
core_biz_area             
0                20.259687
1                32.834660

In [3]:

import numpy as np
# List of file names
files = ['weekly_patterns_2018_sample.csv.zip', 
         'weekly_patterns_2019_sample.csv.zip',
         'weekly_patterns_2020_sample.csv.zip', 
         'weekly_patterns_2021_sample.csv.zip',
         'weekly_patterns_2022_sample.csv.zip']


total_obs = 0
for file in files:
    # Read the data
    data = pd.read_csv(file, compression='zip')
    total_obs += len(data)
print(total_obs)
data.dtypes
applebees_obs = 0
applebees_df = pd.DataFrame() #initializing an empty df

for file in files:
    # Read the data
    data2 = pd.read_csv(file, compression='zip')
    
    # Filter the data for applebees
    data2 = data2.loc[data['brands'].isin(["Applebee's"])]
    
    # Concatenate the data
    applebees_df = pd.concat([applebees_df, data2], ignore_index = True)
    
    applebees_obs = len(applebees_df)
    
print(applebees_obs)
from datetime import datetime, date, timedelta
applebees_df['date_range_start'] = applebees_df['date_range_start'].str[:10]
applebees_df['date_range_end'] = applebees_df['date_range_end'].str[:10]
applebees_df.head(5)

applebees_df['year'] = applebees_df['date_range_start'].str[:4]
applebees_df['date_range_start'] = pd.to_datetime(applebees_df
                                               ['date_range_start'])
applebees_df['date_range_end'] =  pd.to_datetime(applebees_df
                                              ['date_range_end'])
# Split the variable visits_by_day into new variables
applebees_df[['dailyvisits0', 'dailyvisits1', 'dailyvisits2'
           , 'dailyvisits3',
    'dailyvisits4', 'dailyvisits5', 
           'dailyvisits6']] = applebees_df['visits_by_day'].str.split(',', expand=True)
applebees_df['dailyvisits0'] = applebees_df['dailyvisits0'].str[1:3]
applebees_df['dailyvisits6'] = applebees_df['dailyvisits6'].str[0:2]

applebees_df['dailyvisits0'] = applebees_df['dailyvisits0'].astype(float)
applebees_df['dailyvisits1'] = applebees_df['dailyvisits1'].astype(float)
applebees_df['dailyvisits2'] = applebees_df['dailyvisits2'].astype(float)
applebees_df['dailyvisits3'] = applebees_df['dailyvisits3'].astype(float)
applebees_df['dailyvisits4'] = applebees_df['dailyvisits4'].astype(float)
applebees_df['dailyvisits5'] = applebees_df['dailyvisits5'].astype(float)

applebees_df['dailyvisits6'] = applebees_df['dailyvisits6'].str.replace(']', '')

applebees_df['dailyvisits6'] = applebees_df['dailyvisits6'].astype(float)

applebees_df['id'] = applebees_df.index
applebees_df['index'] = applebees_df.index
long_ab = pd.wide_to_long(applebees_df,
                                 ['dailyvisits'
                                 ], i = 'index', j = 'day', sep = '')
long_ab['dailyvisits'] = long_ab['dailyvisits'].astype(int)
aggab = long_ab.groupby(['year']).agg(
{'dailyvisits': 'mean' })

long_ab = long_ab.reset_index()

long_ab['date'] = long_ab['date_range_start'
                                       ] + pd.to_timedelta(long_ab['day'],
                                                           unit='D')

long_ab['dayoftheweek'] = long_ab['date'].dt.day_name()
threshold = 34 #I chose this threshold because 25% of the data will be greater than it: 
                #this gets the top quartile for chilis so need to be comparing same thing

long_ab['manyvisits'] = np.where(long_ab['dailyvisits'] > threshold, 1, 0)
ab_table = pd.crosstab(long_ab['region'] , long_ab['manyvisits'] )
tablepercent = pd.crosstab(long_ab['region'] , long_ab['manyvisits'] ,
                           normalize = 'index')*100
#same table as above but with percentages instead of total number counts
#Done to control for state size and see which states have best performing resturants
long_ab['core_biz_area'] = np.where((long_ab['region'] == 'AL')
                                           | (long_ab['region'] =='AR')
                                           | (long_ab['region'] =='FL') 
                                           | (long_ab['region'] =='GA')
                                           |(long_ab['region'] == 'KS') 
                                           | (long_ab['region'] =='LA')
                                           |(long_ab['region'] == 'MI') 
                                           |(long_ab['region'] == 'MO') 
                                           |(long_ab['region'] == 'MS') 
                                           | (long_ab['region'] =='NC') 
                                           |(long_ab['region'] =='OK') 
                                           |(long_ab['region'] == 'TX') 
                                           |(long_ab['region'] == 'WV'), 1, 0)

long_ab.head()

coreab_agg = long_ab.groupby(['core_biz_area']).agg(
{'dailyvisits': 'mean' })

coreab_agg.head()

6462235
73715


dailyvisits
core_biz_area             
0                15.468671
1                19.656819

In [4]:
long_ab.head(5)

index  day            city region  raw_visit_counts  \
0      0    0      Woodbridge     VA                95   
1      1    0      Marysville     OH               103   
2      2    0       Woodhaven     MI               220   
3      3    0      Shreveport     LA                60   
4      4    0  Crawfordsville     IN               170   

                         safegraph_brand_ids date_range_start  year  id  \
0  SG_BRAND_532a1ea21bbef4dbdc2cae21e89a610a       2018-01-15  2018   0   
1  SG_BRAND_532a1ea21bbef4dbdc2cae21e89a610a       2018-01-08  2018   1   
2  SG_BRAND_532a1ea21bbef4dbdc2cae21e89a610a       2018-02-12  2018   2   
3  SG_BRAND_532a1ea21bbef4dbdc2cae21e89a610a       2018-09-17  2018   3   
4  SG_BRAND_532a1ea21bbef4dbdc2cae21e89a610a       2018-10-15  2018   4   

       brands  postal_code date_range_end  naics_code             placekey  \
0  Applebee's        22193     2018-01-22      722511  222-222@63r-tv4-btv   
1  Applebee's        43040     2018-01-15      722511  222-222@63h-mg6-ht9   
2  Applebee's        48183     2018-02-19      722511  zzw-223@63t-yt6-fs5   
3  Applebee's        71105     2018-09-24      722511  222-222@8f2-d6n-jqf   
4  Applebee's        47933     2018-10-22      722511  222-222@5pv-brf-pqf   

            visits_by_day  dailyvisits       date dayoftheweek  manyvisits  \
0   [14,14,16,13,8,18,12]           14 2018-01-15       Monday           0   
1  [12,12,10,13,11,26,19]           12 2018-01-08       Monday           0   
2  [15,25,27,22,40,54,37]           15 2018-02-12       Monday           0   
3       [2,4,8,15,18,8,5]            2 2018-09-17       Monday           0   
4  [12,21,13,23,40,40,21]           12 2018-10-15       Monday           0   

   core_biz_area  
0              0  
1              0  
2              1  
3              1  
4              0

In [5]:
long_chilis_df.head(5)

index  day            city region  raw_visit_counts  \
0      0    0     Lees Summit     MO               261   
1      1    0  Mount Pleasant     TX               138   
2      2    0          Peoria     IL               111   
3      3    0           Dover     NH               113   
4      4    0          Euless     TX               175   

                         safegraph_brand_ids date_range_start  year  id  \
0  SG_BRAND_072efa4069c7217d82229845c60e6a0d       2018-12-24  2018   0   
1  SG_BRAND_072efa4069c7217d82229845c60e6a0d       2018-08-27  2018   1   
2  SG_BRAND_072efa4069c7217d82229845c60e6a0d       2018-03-19  2018   2   
3  SG_BRAND_072efa4069c7217d82229845c60e6a0d       2018-05-07  2018   3   
4  SG_BRAND_072efa4069c7217d82229845c60e6a0d       2018-02-12  2018   4   

                brands  postal_code date_range_end  naics_code  \
0  Chili's Grill & Bar        64081     2018-12-31      722511   
1  Chili's Grill & Bar        75455     2018-09-03      722511   
2  Chili's Grill & Bar        61614     2018-03-26      722511   
3  Chili's Grill & Bar         3820     2018-05-14      722511   
4  Chili's Grill & Bar        76039     2018-02-19      722511   

              placekey           visits_by_day  dailyvisits       date  \
0  222-222@5pr-4pg-9fz   [26,1,43,45,52,53,41]           26 2018-12-24   
1  223-222@8f2-sqt-kj9  [10,24,14,13,23,24,30]           10 2018-08-27   
2  223-222@5pv-mqc-t5f  [10,10,16,15,21,16,23]           10 2018-03-19   
3  zzw-222@62j-wsq-ysq   [13,15,18,9,26,19,13]           13 2018-05-07   
4  228-222@5qv-x6c-gzf   [26,14,34,9,34,34,24]           26 2018-02-12   

  dayoftheweek  manyvisits  core_biz_area  
0       Monday           0              1  
1       Monday           0              1  
2       Monday           0              0  
3       Monday           0              0  
4       Monday           0              1

The chain I have selected for my benchmark chain is applebees. Applebees and Chilis are 2 pretty similar resturants in terms of amount of locations in the US as well as the overall style of food served at each resturant. As one of Chili's main competetors I would like to see if Applebees saw similar trends as Chili's did and if they are doing better/worse than chili's in certain locations. These could be locations where we can look to move into. Just from running project 1 code for Applebee's it seems they average significantly less visits per day which is an early sign that Chili's is doing better than Applebees nationally. Thus if there are locations where chili's is doing worse, we must look into why that is the case, and propose changes to prevent it. 

In [6]:
applebees_obs = len(long_ab)
print(applebees_obs)

516005


In [7]:
chilis_obs = len(long_chilis_df)
print(chilis_obs)

393694


long_ab is the dataframe for applebees, and long_chilis_df is the same for chilis. Both are in the long form, and have unit place-day. applebees has 516005 observations, which is greater than chili's 393694 observations. This makes sense however because applebees has about 1500 locations in the US whereas chilis only has about 1200

long_ab is my "Benchmark long" and long_chilis_df is my "focal long". Both have identical data structures. The following few blocks of code will be my 'sanity check'

In [8]:
long_chilis_df['dailyvisits'].describe()

count    393694.000000
mean         26.405419
std          27.093758
min           0.000000
25%          13.000000
50%          22.000000
75%          34.000000
max         819.000000
Name: dailyvisits, dtype: float64

In [9]:
long_ab['dailyvisits'].describe()

count    516005.000000
mean         16.952235
std          12.278117
min           0.000000
25%           9.000000
50%          15.000000
75%          23.000000
max         433.000000
Name: dailyvisits, dtype: float64

In [10]:
print('number of chilis observations for each year:')
long_chilis_df['year'].value_counts()

number of chilis observations for each year:


year
2018    79576
2021    78834
2020    78610
2022    78428
2019    78246
Name: count, dtype: int64

In [11]:
print('number of applebees observations for each year:')
long_ab['year'].value_counts()

number of applebees observations for each year:


year
2018    103201
2019    103201
2020    103201
2021    103201
2022    103201
Name: count, dtype: int64

In [12]:
long_ab['placekey'].value_counts()

placekey
223-222@5qv-wx5-2rk    1393
222-222@8fv-zkt-7bk    1372
222-222@63t-mmj-hnq    1372
222-222@8fb-83q-qcq    1365
222-222@63j-rny-qzz    1365
                       ... 
223-222@5qv-x2z-psq      35
225-222@5ps-37f-fxq      28
zzw-226@63g-qrp-94v      21
222-227@8tm-pth-j9z      14
222-222@8gn-23s-d5f       7
Name: count, Length: 945, dtype: int64

In [13]:
long_chilis_df['placekey'].value_counts()

placekey
223-222@5z5-3vw-y35    1834
222-222@8gq-946-r49    1834
222-222@8gb-pwj-wc5    1834
zzw-223@5z6-wdr-fmk    1834
222-222@5pr-3d9-yvz    1834
                       ... 
223-223@63v-46p-z2k      42
23c-222@63s-2xy-835      35
22g-238@63r-bh8-yn5      35
222-223@5sb-85p-m49      35
zzy-222@8t8-cmx-6ff      35
Name: count, Length: 232, dtype: int64

In [14]:
long_ab['placekey'].describe()

count                  516005
unique                    945
top       223-222@5qv-wx5-2rk
freq                     1393
Name: placekey, dtype: object

There are 945 unique applebees and only 232 unique chilis. There is supposed to be more applebees than chilis so this makes sense

In [15]:
long_ab['dayoftheweek'].value_counts()

dayoftheweek
Monday       73715
Tuesday      73715
Wednesday    73715
Thursday     73715
Friday       73715
Saturday     73715
Sunday       73715
Name: count, dtype: int64

In [16]:
long_ab['region'].value_counts()

region
TX    36827
CA    34251
NY    31668
FL    30261
MI    26628
OH    24836
VA    18907
PA    17822
MO    16695
IL    16443
NC    16170
IN    15645
CO    14707
GA    13475
TN    12747
WA    12719
AL    12677
AZ    12278
MD    10976
NJ    10808
MN     9723
SC     9415
NM     8589
WI     8106
MS     8071
KS     7287
OR     5824
CT     5614
NE     5600
IA     5243
MT     4991
ME     4851
ID     4732
WV     4648
LA     4606
MA     4270
KY     4109
DE     3920
OK     3913
NV     3514
UT     2408
SD     2261
ND     2107
RI     1967
AR     1526
NH     1022
AK      882
DC      266
Name: count, dtype: int64

a google search shows that both chains have locations in 50 states, however in the data chilis only has 42 states represented and applebees has 48, however a larger percentage of all applebees stores is includes in the sample so this makes sense. 

In [17]:
focal_sum = long_chilis_df.groupby(['region', 'date']).agg({
    'dailyvisits': ['sum', 'max', 'min', 
                    'median', 'nunique'], 
    'placekey': 'nunique'}).reset_index()
focal_sum.columns = ['state', 'date', 
                     'SumDailyVisits', 'MaxDailyVisitsPerUnit', 
                     'MinDailyVisitsPerUnit', 
                     'MedDailyVisitsPerUnit', 'UniqueVisits', 'Nunits']
focal_sum.head(5)


state       date  SumDailyVisits  MaxDailyVisitsPerUnit  \
0    AL 2018-01-01               7                      7   
1    AL 2018-01-02              10                     10   
2    AL 2018-01-03              18                     18   
3    AL 2018-01-04              14                     14   
4    AL 2018-01-05              30                     30   

   MinDailyVisitsPerUnit  MedDailyVisitsPerUnit  UniqueVisits  Nunits  
0                      7                    7.0             1       1  
1                     10                   10.0             1       1  
2                     18                   18.0             1       1  
3                     14                   14.0             1       1  
4                     30                   30.0             1       1

In [18]:
bench_sum = long_ab.groupby(['region', 'date']).agg({
    'dailyvisits': ['sum', 'max', 'min', 
                    'median', 'nunique'], 
    'placekey': 'nunique'}).reset_index()
bench_sum.columns = ['state', 'date', 
                     'SumDailyVisits', 'MaxDailyVisitsPerUnit', 
                     'MinDailyVisitsPerUnit', 
                     'MedDailyVisitsPerUnit', 'UniqueVisits', 'Nunits']
bench_sum.head(5)

state       date  SumDailyVisits  MaxDailyVisitsPerUnit  \
0    AK 2018-01-22              16                     16   
1    AK 2018-01-23               9                      9   
2    AK 2018-01-24               8                      8   
3    AK 2018-01-25              13                     13   
4    AK 2018-01-26              22                     22   

   MinDailyVisitsPerUnit  MedDailyVisitsPerUnit  UniqueVisits  Nunits  
0                     16                   16.0             1       1  
1                      9                    9.0             1       1  
2                      8                    8.0             1       1  
3                     13                   13.0             1       1  
4                     22                   22.0             1       1

In [19]:
print("Number of observations in Benchmark Summary:", len(bench_sum))

print("Number of observations in Focal Summary:", len(focal_sum))

Number of observations in Benchmark Summary: 80668
Number of observations in Focal Summary: 74788


The benchmark has more observations, this makes sense because applebees is represented in 6 more states in the dataset as it is data from 48 whereas chilis only has data from 42. 

In [20]:
MergedSummary = pd.merge(bench_sum, focal_sum, on=['state', 'date'], how='inner')

MergedSummary.head()

state       date  SumDailyVisits_x  MaxDailyVisitsPerUnit_x  \
0    AL 2018-01-01                68                       20   
1    AL 2018-01-02                39                        9   
2    AL 2018-01-03                70                       25   
3    AL 2018-01-04                73                       25   
4    AL 2018-01-05               106                       35   

   MinDailyVisitsPerUnit_x  MedDailyVisitsPerUnit_x  UniqueVisits_x  Nunits_x  \
0                        3                      8.0               7         7   
1                        3                      5.0               6         7   
2                        2                      9.0               6         7   
3                        5                      9.0               5         7   
4                        5                     15.0               6         7   

   SumDailyVisits_y  MaxDailyVisitsPerUnit_y  MinDailyVisitsPerUnit_y  \
0                 7                        7                        7   
1                10                       10                       10   
2                18                       18                       18   
3                14                       14                       14   
4                30                       30                       30   

   MedDailyVisitsPerUnit_y  UniqueVisits_y  Nunits_y  
0                      7.0               1         1  
1                     10.0               1         1  
2                     18.0               1         1  
3                     14.0               1         1  
4                     30.0               1         1

In [21]:
# Number of matched observations
matched = MergedSummary.shape[0]

# Number of observations in BenchmarkSummary but not in FocalSummary
unmatched_bench = bench_sum.shape[0] - matched

# Number of observations in FocalSummary but not in BenchmarkSummary
unmatched_focal = focal_sum.shape[0] - matched

In [22]:
print(matched)
print(unmatched_bench)
print(unmatched_focal)

67151
13517
7637


In [23]:
# Get the unique states from both datasets
states_bench = set(bench_sum['state'].unique())
states_focal = set(focal_sum['state'].unique())

# Find the common states
common_states = states_bench & states_focal

# Filter the merged dataset to only keep the common states
MergedSummary = MergedSummary[MergedSummary['state'].isin(common_states)]

In [24]:
print(common_states)

{'AL', 'TN', 'OH', 'IL', 'MO', 'NM', 'IN', 'PA', 'MD', 'WI', 'KS', 'WV', 'MN', 'IA', 'NV', 'CO', 'DE', 'FL', 'MI', 'KY', 'OK', 'AR', 'AZ', 'NY', 'NJ', 'UT', 'WA', 'RI', 'CT', 'MT', 'NH', 'MA', 'GA', 'ID', 'CA', 'LA', 'NC', 'MS', 'TX', 'VA'}


In [25]:
print(states_bench)
print(states_focal)

{'AL', 'TN', 'DC', 'OH', 'IL', 'MO', 'NM', 'IN', 'PA', 'MD', 'WI', 'KS', 'WV', 'IA', 'MN', 'SD', 'DE', 'CO', 'NV', 'FL', 'MI', 'KY', 'OK', 'ND', 'AR', 'AK', 'AZ', 'NY', 'ME', 'NJ', 'UT', 'WA', 'RI', 'CT', 'MT', 'NH', 'MA', 'NE', 'GA', 'ID', 'CA', 'OR', 'LA', 'NC', 'MS', 'TX', 'VA', 'SC'}
{'AL', 'TN', 'OH', 'IL', 'MO', 'NM', 'IN', 'PA', 'MD', 'WI', 'KS', 'WV', 'IA', 'MN', 'DE', 'CO', 'NV', 'FL', 'HI', 'MI', 'KY', 'OK', 'VT', 'AR', 'AZ', 'NY', 'NJ', 'UT', 'WA', 'RI', 'CT', 'MT', 'NH', 'MA', 'GA', 'ID', 'CA', 'LA', 'NC', 'MS', 'TX', 'VA'}


In [26]:
print(states_focal - states_bench)

{'VT', 'HI'}


In [27]:
print(states_bench - states_focal)

{'AK', 'DC', 'NE', 'OR', 'ME', 'SC', 'ND', 'SD'}


There are 40 matched states, the 2 from focal that are not in the benchmark are Vermont (VT) and Hawaii (HI), the 8  missing from focal that are in benchmark are: {'NE', 'AK', 'ND', 'ME', 'OR', 'SD', 'SC', 'DC'}

The reasoning for the imperfect matching is due to the fact that this is not a full sample of all resturants, but a subset that has been randomly sampled. Additionally these 10 states are states where one or both of the chains do not have many locations, so it is natural to be missing some states and have some asymmetry in the data. 

In [28]:
MergedSummary.head()

state       date  SumDailyVisits_x  MaxDailyVisitsPerUnit_x  \
0    AL 2018-01-01                68                       20   
1    AL 2018-01-02                39                        9   
2    AL 2018-01-03                70                       25   
3    AL 2018-01-04                73                       25   
4    AL 2018-01-05               106                       35   

   MinDailyVisitsPerUnit_x  MedDailyVisitsPerUnit_x  UniqueVisits_x  Nunits_x  \
0                        3                      8.0               7         7   
1                        3                      5.0               6         7   
2                        2                      9.0               6         7   
3                        5                      9.0               5         7   
4                        5                     15.0               6         7   

   SumDailyVisits_y  MaxDailyVisitsPerUnit_y  MinDailyVisitsPerUnit_y  \
0                 7                        7                        7   
1                10                       10                       10   
2                18                       18                       18   
3                14                       14                       14   
4                30                       30                       30   

   MedDailyVisitsPerUnit_y  UniqueVisits_y  Nunits_y  
0                      7.0               1         1  
1                     10.0               1         1  
2                     18.0               1         1  
3                     14.0               1         1  
4                     30.0               1         1

In [29]:
MergedSummary['AvgDailyVisits_x'] = MergedSummary['SumDailyVisits_x'] / MergedSummary['Nunits_x']
MergedSummary['AvgDailyVisits_y'] = MergedSummary['SumDailyVisits_y'] / MergedSummary['Nunits_y']

In [30]:
MergedSummary.head()

state       date  SumDailyVisits_x  MaxDailyVisitsPerUnit_x  \
0    AL 2018-01-01                68                       20   
1    AL 2018-01-02                39                        9   
2    AL 2018-01-03                70                       25   
3    AL 2018-01-04                73                       25   
4    AL 2018-01-05               106                       35   

   MinDailyVisitsPerUnit_x  MedDailyVisitsPerUnit_x  UniqueVisits_x  Nunits_x  \
0                        3                      8.0               7         7   
1                        3                      5.0               6         7   
2                        2                      9.0               6         7   
3                        5                      9.0               5         7   
4                        5                     15.0               6         7   

   SumDailyVisits_y  MaxDailyVisitsPerUnit_y  MinDailyVisitsPerUnit_y  \
0                 7                        7                        7   
1                10                       10                       10   
2                18                       18                       18   
3                14                       14                       14   
4                30                       30                       30   

   MedDailyVisitsPerUnit_y  UniqueVisits_y  Nunits_y  AvgDailyVisits_x  \
0                      7.0               1         1          9.714286   
1                     10.0               1         1          5.571429   
2                     18.0               1         1         10.000000   
3                     14.0               1         1         10.428571   
4                     30.0               1         1         15.142857   

   AvgDailyVisits_y  
0               7.0  
1              10.0  
2              18.0  
3              14.0  
4              30.0

In [31]:
average_visits_df = MergedSummary[['AvgDailyVisits_x', 'AvgDailyVisits_y']]
average_visits_df.head()

AvgDailyVisits_x  AvgDailyVisits_y
0          9.714286               7.0
1          5.571429              10.0
2         10.000000              18.0
3         10.428571              14.0
4         15.142857              30.0

In [32]:
average_visits_applebees = MergedSummary['SumDailyVisits_x'].mean()
average_visits_chilis = MergedSummary['SumDailyVisits_y'].mean()
Average_stores_applebees = MergedSummary['Nunits_x'].mean()
Average_stores_chilis = MergedSummary['Nunits_y'].mean()

# Create a DataFrame to display the results
average_visits_df = pd.DataFrame({
    'Chain': ['applebees', 'chilis'],
    'Average SumDailyVisits': [average_visits_applebees, average_visits_chilis],
    'Average locations' : [Average_stores_applebees , Average_stores_chilis]
})
average_visits_df.head()

Chain  Average SumDailyVisits  Average locations
0  applebees              120.956010           7.094861
1     chilis              148.820166           5.647451

From the table, accross all states chilis seems to average almost 30 more customers per day with less locations So it seems, at least from a very high up perspective that Chilis is outperforming its competitor applebees  

In [33]:
import scipy
print(scipy.__version__)

1.11.1


In [34]:
from scipy.stats import ttest_ind

In [35]:
import scipy.stats as stats

# Calculate the t-statistic and p-value
t_statistic, p_value = stats.ttest_ind(MergedSummary['SumDailyVisits_x'],
                                       MergedSummary['SumDailyVisits_y'], equal_var=True)

# Define the significance level (alpha)
alpha = 0.05

print(t_statistic)
print(p_value)
print('Null Hypothesis:no difference between The average sum of dailyvisits between Chilis and Applebees')

# Check if the p-value is less than alpha
if p_value < alpha:
    result = "Reject the null hypothesis "
else:
    result = "Fail to reject the null hypothesis "

result


-27.53908826728215
1.7423039244206804e-166
Null Hypothesis: There is no difference between The average sum of dailyvisits between Chilis and Applebees


'Reject the null hypothesis '

With the P value being essentially 0, we reject the null hypothesis that there is no difference between the avg sum of daily visits. This confirms my assumption from the previous question that I determined by just looking at the numbers

In [36]:
MergedSummary['Post03132020'] = (MergedSummary[
    'date'] >= '2020-03-13').astype(int)


In [37]:
MergedSummary.head()

state       date  SumDailyVisits_x  MaxDailyVisitsPerUnit_x  \
0    AL 2018-01-01                68                       20   
1    AL 2018-01-02                39                        9   
2    AL 2018-01-03                70                       25   
3    AL 2018-01-04                73                       25   
4    AL 2018-01-05               106                       35   

   MinDailyVisitsPerUnit_x  MedDailyVisitsPerUnit_x  UniqueVisits_x  Nunits_x  \
0                        3                      8.0               7         7   
1                        3                      5.0               6         7   
2                        2                      9.0               6         7   
3                        5                      9.0               5         7   
4                        5                     15.0               6         7   

   SumDailyVisits_y  MaxDailyVisitsPerUnit_y  MinDailyVisitsPerUnit_y  \
0                 7                        7                        7   
1                10                       10                       10   
2                18                       18                       18   
3                14                       14                       14   
4                30                       30                       30   

   MedDailyVisitsPerUnit_y  UniqueVisits_y  Nunits_y  AvgDailyVisits_x  \
0                      7.0               1         1          9.714286   
1                     10.0               1         1          5.571429   
2                     18.0               1         1         10.000000   
3                     14.0               1         1         10.428571   
4                     30.0               1         1         15.142857   

   AvgDailyVisits_y  Post03132020  
0               7.0             0  
1              10.0             0  
2              18.0             0  
3              14.0             0  
4              30.0             0

In [38]:
blue_states = ['ME', 'NH', 'MA', 'CT', 'NY', 'RI', 'NJ', 
               'MD', 'VA', 'IL', 'MN', 'CO', 'NM', 'NV', 
               'CA', 'OR', 'WA']
MergedSummary['BlueState'] = MergedSummary['state'].isin(blue_states).astype(int)

In [39]:
MergedSummary.head()


state       date  SumDailyVisits_x  MaxDailyVisitsPerUnit_x  \
0    AL 2018-01-01                68                       20   
1    AL 2018-01-02                39                        9   
2    AL 2018-01-03                70                       25   
3    AL 2018-01-04                73                       25   
4    AL 2018-01-05               106                       35   

   MinDailyVisitsPerUnit_x  MedDailyVisitsPerUnit_x  UniqueVisits_x  Nunits_x  \
0                        3                      8.0               7         7   
1                        3                      5.0               6         7   
2                        2                      9.0               6         7   
3                        5                      9.0               5         7   
4                        5                     15.0               6         7   

   SumDailyVisits_y  MaxDailyVisitsPerUnit_y  MinDailyVisitsPerUnit_y  \
0                 7                        7                        7   
1                10                       10                       10   
2                18                       18                       18   
3                14                       14                       14   
4                30                       30                       30   

   MedDailyVisitsPerUnit_y  UniqueVisits_y  Nunits_y  AvgDailyVisits_x  \
0                      7.0               1         1          9.714286   
1                     10.0               1         1          5.571429   
2                     18.0               1         1         10.000000   
3                     14.0               1         1         10.428571   
4                     30.0               1         1         15.142857   

   AvgDailyVisits_y  Post03132020  BlueState  
0               7.0             0          0  
1              10.0             0          0  
2              18.0             0          0  
3              14.0             0          0  
4              30.0             0          0

In [40]:
MergedSummary['BluePost'] = MergedSummary['BlueState'] * MergedSummary['Post03132020']
MergedSummary.head()

state       date  SumDailyVisits_x  MaxDailyVisitsPerUnit_x  \
0    AL 2018-01-01                68                       20   
1    AL 2018-01-02                39                        9   
2    AL 2018-01-03                70                       25   
3    AL 2018-01-04                73                       25   
4    AL 2018-01-05               106                       35   

   MinDailyVisitsPerUnit_x  MedDailyVisitsPerUnit_x  UniqueVisits_x  Nunits_x  \
0                        3                      8.0               7         7   
1                        3                      5.0               6         7   
2                        2                      9.0               6         7   
3                        5                      9.0               5         7   
4                        5                     15.0               6         7   

   SumDailyVisits_y  MaxDailyVisitsPerUnit_y  MinDailyVisitsPerUnit_y  \
0                 7                        7                        7   
1                10                       10                       10   
2                18                       18                       18   
3                14                       14                       14   
4                30                       30                       30   

   MedDailyVisitsPerUnit_y  UniqueVisits_y  Nunits_y  AvgDailyVisits_x  \
0                      7.0               1         1          9.714286   
1                     10.0               1         1          5.571429   
2                     18.0               1         1         10.000000   
3                     14.0               1         1         10.428571   
4                     30.0               1         1         15.142857   

   AvgDailyVisits_y  Post03132020  BlueState  BluePost  
0               7.0             0          0         0  
1              10.0             0          0         0  
2              18.0             0          0         0  
3              14.0             0          0         0  
4              30.0             0          0         0

In [41]:
MergedSummary['BluePost'].value_counts()

BluePost
0    53598
1    13553
Name: count, dtype: int64

In [42]:
import statsmodels.formula.api as smf

In [43]:
def log_plus_1(x):
    return np.log(x + 1)


In [44]:
results = smf.ols('log_plus_1(SumDailyVisits_y) ~ C(Post03132020) + C(BlueState) + C(BluePost)',
                  data=MergedSummary).fit()


results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                 
========================================================================================
Dep. Variable:     log_plus_1(SumDailyVisits_y)   R-squared:                       0.029
Model:                                      OLS   Adj. R-squared:                  0.029
Method:                           Least Squares   F-statistic:                     665.9
Date:                          Sun, 31 Mar 2024   Prob (F-statistic):               0.00
Time:                                  15:06:02   Log-Likelihood:            -1.1779e+05
No. Observations:                         67151   AIC:                         2.356e+05
Df Residuals:                             67147   BIC:                         2.356e+05
Df Model:                                     3                                         
Covariance Type:                      nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
Intercept                4.3831      0.010    432.901      0.000       4.363       4.403
C(Post03132020)[T.1]    -0.0948      0.014     -6.992      0.000      -0.121      -0.068
C(BlueState)[T.1]       -0.4851      0.017    -28.532      0.000      -0.518      -0.452
C(BluePost)[T.1]        -0.0120      0.023     -0.528      0.597      -0.056       0.032
==============================================================================
Omnibus:                     2425.376   Durbin-Watson:                   0.149
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2719.344
Skew:                          -0.468   Prob(JB):                         0.00
Kurtosis:                       3.309   Cond. No.                         6.74
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The coefficients on Post03132020 & BlueState are both statistically significant at the 95% level as seen by the P values being 0. The coefficient on BluePost however is not significant at the 95 % level with a P value of almost .6, & 0 being included in the 95% confidence Interval

Since we are in the log-normal form, the three coefficients mean that: 1.) being after the national emergency decreased the sum of visits per day by 9.5%. 2.) Being a blue state decreases the sum of visits perday by 48.5% & 3.) being a blue state post pandemic has an expected 1.2% decrease in sum daily visits. 

Thus I would accept the null hypothesis that The national emergency announced on March 13, 2020 does not make a difference on average SumDailyVisits of your own chain between red and blue states”? 

The national emergency itself definitely had an effect, but Blue and red states saw similar percentage decreases in dailyvisits 

In [45]:
 !pip install linearmodels

In [46]:
from linearmodels import PanelOLS

In [48]:
df2 = MergedSummary.copy()
df2 = MergedSummary.set_index(['state', 'date'])
df2.head()

SumDailyVisits_x  MaxDailyVisitsPerUnit_x  \
state date                                                    
AL    2018-01-01                68                       20   
      2018-01-02                39                        9   
      2018-01-03                70                       25   
      2018-01-04                73                       25   
      2018-01-05               106                       35   

                  MinDailyVisitsPerUnit_x  MedDailyVisitsPerUnit_x  \
state date                                                           
AL    2018-01-01                        3                      8.0   
      2018-01-02                        3                      5.0   
      2018-01-03                        2                      9.0   
      2018-01-04                        5                      9.0   
      2018-01-05                        5                     15.0   

                  UniqueVisits_x  Nunits_x  SumDailyVisits_y  \
state date                                                     
AL    2018-01-01               7         7                 7   
      2018-01-02               6         7                10   
      2018-01-03               6         7                18   
      2018-01-04               5         7                14   
      2018-01-05               6         7                30   

                  MaxDailyVisitsPerUnit_y  MinDailyVisitsPerUnit_y  \
state date                                                           
AL    2018-01-01                        7                        7   
      2018-01-02                       10                       10   
      2018-01-03                       18                       18   
      2018-01-04                       14                       14   
      2018-01-05                       30                       30   

                  MedDailyVisitsPerUnit_y  UniqueVisits_y  Nunits_y  \
state date                                                            
AL    2018-01-01                      7.0               1         1   
      2018-01-02                     10.0               1         1   
      2018-01-03                     18.0               1         1   
      2018-01-04                     14.0               1         1   
      2018-01-05                     30.0               1         1   

                  AvgDailyVisits_x  AvgDailyVisits_y  Post03132020  BlueState  \
state date                                                                      
AL    2018-01-01          9.714286               7.0             0          0   
      2018-01-02          5.571429              10.0             0          0   
      2018-01-03         10.000000              18.0             0          0   
      2018-01-04         10.428571              14.0             0          0   
      2018-01-05         15.142857              30.0             0          0   

                  BluePost  
state date                  
AL    2018-01-01         0  
      2018-01-02         0  
      2018-01-03         0  
      2018-01-04         0  
      2018-01-05         0

In [49]:
df2[['BlueState', 'Post03132020', 'BluePost']].head()

BlueState  Post03132020  BluePost
state date                                         
AL    2018-01-01          0             0         0
      2018-01-02          0             0         0
      2018-01-03          0             0         0
      2018-01-04          0             0         0
      2018-01-05          0             0         0

In [61]:
mod2 = PanelOLS(log_plus_1(df2['SumDailyVisits_y']), 
               df2[['BlueState', 'Post03132020', 'BluePost']], 
               entity_effects = True,
               drop_absorbed = True)
results2 = mod2.fit(cov_type='clustered')
print(results2)

C:\Users\mattm\AppData\Local\Temp\ipykernel_41080\2351371437.py:5: AbsorbingEffectWarning: 
Variables have been fully absorbed and have removed from the regression:

BlueState

  results2 = mod2.fit(cov_type='clustered')


                          PanelOLS Estimation Summary                           
Dep. Variable:       SumDailyVisits_y   R-squared:                        0.0094
Estimator:                   PanelOLS   R-squared (Between):             -0.0274
No. Observations:               67151   R-squared (Within):               0.0094
Date:                Sun, Mar 31 2024   R-squared (Overall):             -0.0264
Time:                        15:44:12   Log-likelihood                -5.997e+04
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      318.85
Entities:                          40   P-value                           0.0000
Avg Obs:                       1678.8   Distribution:                 F(2,67109)
Min Obs:                       406.00                                           
Max Obs:                       1827.0   F-statistic (robust):             350.89
                            

The above code is for just State FE. Thus BlueState gets absorbed. Interestingly, now BluePost has a statistically significant coefficient suggesting that blue states did lead to a decrease of 7.47% of avg daily visits. 

In [62]:
mod = PanelOLS(log_plus_1(df2['SumDailyVisits_y']), 
               df2[['BlueState', 'Post03132020', 'BluePost']], 
               time_effects = True,
               drop_absorbed = True)
results = mod.fit(cov_type='clustered')
print(results)

C:\Users\mattm\AppData\Local\Temp\ipykernel_41080\1014797941.py:5: AbsorbingEffectWarning: 
Variables have been fully absorbed and have removed from the regression:

Post03132020

  results = mod.fit(cov_type='clustered')


                          PanelOLS Estimation Summary                           
Dep. Variable:       SumDailyVisits_y   R-squared:                        0.0296
Estimator:                   PanelOLS   R-squared (Between):             -0.0771
No. Observations:               67151   R-squared (Within):               0.0020
Date:                Sun, Mar 31 2024   R-squared (Overall):             -0.0739
Time:                        15:44:18   Log-likelihood                -1.148e+05
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      995.54
Entities:                          40   P-value                           0.0000
Avg Obs:                       1678.8   Distribution:                 F(2,65322)
Min Obs:                       406.00                                           
Max Obs:                       1827.0   F-statistic (robust):             1087.5
                            

Just including the time FE Guves a super high coefficient for BlueState, but this is because as seen in the previous data, chilis core business area is mostly red states like Texas, Alabama, and Florida. so without including state FE this number is inflated

In [58]:
mod = PanelOLS(log_plus_1(df2['SumDailyVisits_y']), 
               df2[['BlueState', 'Post03132020', 'BluePost']], 
               time_effects = True, entity_effects = True,
               drop_absorbed = True)
results = mod.fit(cov_type='clustered')
print(results)

C:\Users\mattm\AppData\Local\Temp\ipykernel_41080\2366714972.py:5: AbsorbingEffectWarning: 
Variables have been fully absorbed and have removed from the regression:

BlueState, Post03132020

  results = mod.fit(cov_type='clustered')


                          PanelOLS Estimation Summary                           
Dep. Variable:       AvgDailyVisits_y   R-squared:                        0.0030
Estimator:                   PanelOLS   R-squared (Between):             -0.0118
No. Observations:               67151   R-squared (Within):               0.0056
Date:                Sun, Mar 31 2024   R-squared (Overall):             -0.0109
Time:                        15:24:41   Log-likelihood                 -3.51e+04
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      198.72
Entities:                          40   P-value                           0.0000
Avg Obs:                       1678.8   Distribution:                 F(1,65284)
Min Obs:                       406.00                                           
Max Obs:                       1827.0   F-statistic (robust):             217.28
                            

Including both effects leaves us with just the BluePost variable, and the coefficient is statistically significant suggesting differently that the original regression from part a that being a bluestate post 03132020 does in fact decrease average daily visits by 9.5 percent

Including both Fixed effects is definately better for hypothesis testing as it allows us to see through state by state differences and get a better, and more importantly statistically significant coefficient. Thus using the Fixed effects we can reject the aforementioned hypothesis whereas without adding the fixed effects we could not reject. 

In [60]:
difference = log_plus_1(MergedSummary['SumDailyVisits_y']) - log_plus_1(MergedSummary['SumDailyVisits_x'])

results2 = smf.ols('difference ~ C(Post03132020) + C(BlueState) + C(BluePost)',
                   data=MergedSummary).fit()


results2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             difference   R-squared:                       0.018
Model:                            OLS   Adj. R-squared:                  0.018
Method:                 Least Squares   F-statistic:                     412.6
Date:                Sun, 31 Mar 2024   Prob (F-statistic):          1.31e-265
Time:                        15:36:01   Log-Likelihood:            -1.0664e+05
No. Observations:               67151   AIC:                         2.133e+05
Df Residuals:                   67147   BIC:                         2.133e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
Intercept               -0.0097      0.009     -1.133      0.257      -0.027       0.007
C(Post03132020)[T.1]    -0.1195      0.011    -10.411      0.000      -0.142      -0.097
C(BlueState)[T.1]       -0.4865      0.014    -33.783      0.000      -0.515      -0.458
C(BluePost)[T.1]         0.3620      0.019     18.843      0.000       0.324       0.400
==============================================================================
Omnibus:                     2425.873   Durbin-Watson:                   0.193
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3249.340
Skew:                          -0.388   Prob(JB):                         0.00
Kurtosis:                       3.747   Cond. No.                         6.74
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The Coefficients on Post03132020 and BlueState are still very similar to the previous regression. However, BluePost now has a strong positive Coefficient. This suggests that Chilis in blue states greatly outperformed applebees in blue states after the pandemic. 

In [69]:
df2['difference'] = log_plus_1(df2['SumDailyVisits_y']) - log_plus_1(df2['SumDailyVisits_x'])

In [70]:
mod = PanelOLS(df2['difference'], 
               df2[['BlueState', 'Post03132020', 'BluePost']], 
               time_effects = True, entity_effects = True,
               drop_absorbed = True)
results = mod.fit(cov_type='clustered')
print(results)

C:\Users\mattm\AppData\Local\Temp\ipykernel_41080\838643815.py:5: AbsorbingEffectWarning: 
Variables have been fully absorbed and have removed from the regression:

BlueState, Post03132020

  results = mod.fit(cov_type='clustered')


                          PanelOLS Estimation Summary                           
Dep. Variable:             difference   R-squared:                        0.0047
Estimator:                   PanelOLS   R-squared (Between):             -0.0381
No. Observations:               67151   R-squared (Within):               0.0008
Date:                Sun, Mar 31 2024   R-squared (Overall):             -0.0204
Time:                        15:52:56   Log-likelihood                -7.278e+04
Cov. Estimator:             Clustered                                           
                                        F-statistic:                      305.05
Entities:                          40   P-value                           0.0000
Avg Obs:                       1678.8   Distribution:                 F(1,65284)
Min Obs:                       406.00                                           
Max Obs:                       1827.0   F-statistic (robust):             325.61
                            

even in adding state and time fixed effects, we still see a strong, positive, statistically significant coefficient on BluePost. 

This change could be due to the fact that generally, Blue states put in place longer & more extreme social distancing practices. So from this data it would seem as though it was harder for resturants to operate in blue states than red states after the pandemic. By using the difference as the dependent variable we can see which chain adapted better to the extreme social distancing measures, and it appears as though Chili's did so much better than applebees. I know that chilis greatly accomodated takeout availability during covid, which could be one explanation. 